### PR TITLE
Doc: Reorganize math module documentation

### PR DESCRIPTION
**Pull request by Nodd**
Following #125810, here is a proposal to reorganize the `math` documentation. I took into account the comments from the previous PRs : https://github.com/python/cpython/pull/125810#discussion_r1814366833 https://github.com/python/cpython/pull/125810#pullrequestreview-2409299006 https://github.com/python/cpython/pull/125810#discussion_r1814370755 and tried to improve from here. As suggested in a comment, I used https://en.cppreference.com/w/c/numeric/math as reference.

One thing I'm not fond of is the duplication of the paragraph about the return values, now that `frexp` and `modf` are in different sections. Note that `modf` is also cited in another paragraph in *Basic floating point operations*.

Another question I have is, should the functions be listed in alphabetic order, or in a logical order ? For example, the trigonometric functions could start with `sin` and `cos` instead of `acos` and `asin`, or `gcd` and `lcm` could be listed together. Same with `sqrt` and `cbrt`.

[Direct link to the updated `math` module documentation](https://cpython-previews--126337.org.readthedocs.build/en/126337/library/math.html)

----
📚 Documentation preview 📚: https://cpython-previews--126337.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/math.rst
Comment (picnixz) [inline Doc/library/math.rst]:
```suggestion
**Floating point arithmetic**
```

File: Doc/library/math.rst
Comment (picnixz) [inline Doc/library/math.rst]: I think it's fine to keep "Special functions" because in mathematics we call them "Special functions": https://en.wikipedia.org/wiki/Special_functions. And this wouldn't limit us to adding more special functions.

**Review by picnixz — COMMENTED** [review]
cc @skirpichev

File: Doc/library/math.rst
Comment (Nodd) [reply Doc/library/math.rst]: I didn't know that, thank you for the information.

**Comment by Nodd** [comment]
Thanks @picnixz for the review, I included your modifications.

File: Doc/library/math.rst
Comment (picnixz) [reply Doc/library/math.rst]: No worries! I know that the C++ reference uses Error & Gamma functions but I think Python decided to go for the more general category of functions.

File: Doc/library/math.rst
Comment (skirpichev) [inline Doc/library/math.rst]:
```suggestion
:func:`floor(x)  <floor>`                             Floor of *x*, the largest integer less than or equal to *x*
```

File: Doc/library/math.rst
Comment (skirpichev) [inline Doc/library/math.rst:39]: I think you should sync table names with actual section headers.

File: Doc/library/math.rst
Comment (skirpichev) [reply Doc/library/math.rst]: I don't think we will add anything beyond C standard.

File: Doc/library/math.rst
Comment (Nodd) [reply Doc/library/math.rst:39]: Of course! I forgot to update the section header, it's done now.

File: Doc/library/math.rst
Comment (Nodd) [reply Doc/library/math.rst]: "Error and gamma functions" is the title used in the C documentation page. I initially changed it because I found it more explicit, but I didn't know the mathematical meaning of "Special functions". Both seem fine, so maybe status quo wins?

**Comment by vstinner** [comment]
Overall, I like the new math doc organization. Nice work.

> Special functions

I'm not sure how special are these functions. Maybe just say "Other functions"?

**Comment by picnixz** [comment]
We had a brief discussion on this in https://github.com/python/cpython/pull/126337#discussion_r1826679366 (see also https://mathworld.wolfram.com/SpecialFunction.html).

Now, not everyone knows about the "special functions" terminology which is first informal (there is no formal definition of what a "special function", it's just that we grouped some functions under the category of "Special functions") and second, it allows us to add whatever miscelleanous functions we want to add (although Sergey hinted that we probably won't add more than what the C standard does). So I think "Miscellaneous functions" or "Other functions" is also a good title.

**Comment by Nodd** [comment]
@vstinner Thank you for you kind words.

About special functions, I proposed a change, see [this comment thread](https://github.com/python/cpython/pull/126337#discussion_r1826679366). I learned that those are actually [special](https://en.wikipedia.org/wiki/Special_functions).

**Comment by Nodd** [comment]
After thinking a bit more about it, one advantage of "Special function" is that it has a meaning to the specialists, and is no less informative than "Miscellaneous functions" or "Other functions" to other people. Those titles are too generic to actually mean anything.

My initial proposal was to be explicit by using "Error and gamma functions". This title could be changed if another function is added in the future. In the end, I simply went with status quo wins.

**Comment by skirpichev** [comment]
> My initial proposal was to be explicit by using "Error and gamma functions".

This is a best variant, IMO.  It's aligned with the C standard.

More generic title (like "Special functions") opens also the door for enhancement requests "lets add also this and that special functions".

File: Doc/library/math.rst
Comment (skirpichev) [inline Doc/library/math.rst]:
```suggestion
   Return ``fmod(x, y)``, as defined by the platform C library. Note that the
```

This seems to be unrelated to current pr.  Lets just move functions around.

File: Doc/library/math.rst
Comment (skirpichev) [inline Doc/library/math.rst]: I think this either should be (1) moved to modf sphinx docs or (2) just removed.  I would prefer (2).

File: Doc/library/math.rst
Comment (skirpichev) [inline Doc/library/math.rst]: Same as for note about modf.

File: Doc/library/math.rst
Comment (vstinner) [reply Doc/library/math.rst]: Please don't remove content in a big refactoring PR. Create a separated PR just for that.

File: Doc/library/math.rst
Comment (Nodd) [reply Doc/library/math.rst]: I agree, I created issue #126719 to track this modification. I will remove the modification from this PR.

File: Doc/library/math.rst
Comment (Nodd) [reply Doc/library/math.rst]: I created an issue to track the suggestion: #126722

Maybe I should move the paragraph to the `modf` documentation in the meantime?

Comment (skirpichev) [reply Doc/library/math.rst]: The rationale to merge this with the modf docs is that a separate note just for one function is less readable.  Before, such note covered *several* functions, now they are in different sections.

Comment (Nodd) [reply Doc/library/math.rst]: I moved the notes nor now.

File: Doc/library/math.rst
Comment (Nodd) [reply Doc/library/math.rst]: I moved the notes nor now. See https://github.com/python/cpython/issues/126722 for later.

**Comment by Nodd** [comment]
I resolved all the conversations.

@picnixz @skirpichev There is only the remaining question about "Special functions" vs "Error and gamma functions" vs something else. Who gets to decide?

**Review by vstinner — APPROVED** [review]
LGTM.

I like the new categories. The change mostly moves functions, the few extra changes are acceptable (I recreated the PR manually to detect "extra" changes").

I'm fine with "Special functions" which is already used.

File: Doc/library/math.rst
Comment (skirpichev) [inline Doc/library/math.rst]: One last nitpick.  Lets not move this footer.

Edit: I meant the whole footer, not just the reference to cmath.

**Review by picnixz — APPROVED** [review]
LGTM (modulo Sergey's comment).

File: Doc/library/math.rst
Comment (Nodd) [reply Doc/library/math.rst]: I had two things in mind when moving this block:
* I find it not easily discoverable, I wasn't expecting another general paragraph *after* the API listing
* In the page layout, it effectively appears in the `Constants` section, which I find confusing.

I propose to put it back at the end as you asked, and to create a new section named `Implementation detail`. That would fix both of my concerns.

Comment (skirpichev) [reply Doc/library/math.rst]: > I find it not easily discoverable

It a good thing.

> In the page layout, it effectively appears in the Constants section, which I find confusing.

I think that it's visually distinct from the sphinx docs of constants.  No need for extra section.

Comment (Nodd) [reply Doc/library/math.rst]: Alright, the block is back to the end of the file.

**Review by skirpichev — APPROVED** [review]

**Comment by vstinner** [comment]
Merged, thank you.

Should we backport the change to avoid merge conflicts?

**Comment by skirpichev** [comment]
> Should we backport the change to avoid merge conflicts?

Yes, I think so.  Because #125810 was backported.

**Comment by vstinner** [comment]
@Nodd: Would you mind to try to backport this change to the 3.13 branch manually? There are merge conflicts.

**Comment by skirpichev** [comment]
3.12 backport also fails.  Its all coming from https://github.com/python/cpython/pull/126215.

**Comment by Nodd** [comment]
Weird, I thought I already rebased the PR after the other modification. I'll look into it.

**Comment by vstinner** [comment]
> Weird, I thought I already rebased the PR after the other modification. I'll look into it.

The change is merged into the main branch. I'm asking about backporting the change to the 3.13 branch. Maybe some other doc changes were not backported to 3.13?

**Comment by skirpichev** [comment]
I took liberty to do backport for 3.13.

The 3.12 backport blocked by https://github.com/python/cpython/pull/126309.

**Comment by Nodd** [comment]
Thank you @skirpichev for the backport.

**Pull request merged**
